### DISCUSSION TITLE
Does UNO Platform support xaml island?

### DISCUSSION THREAD

**Discussion by samsunlu55atakum** [unoplatform/uno#13549] [Q&A] [answered]
I am thinking of porting my wpf app to uno platform. But firstly, i need to convert to uwp. So, i am using xaml island to use uwp controls in wpf. But, Can I move my wpf application that I use uwp controls via xaml island to uno and run it on linux or macOS (cross platform)?

**Comment by jeromelaban** [comment]
Thanks for your interesting in Uno! At this time, [Uno Islands](https://platform.uno/docs/articles/guides/uno-islands.html) is a Windows-only feature, as it allows to run Uno code inside of a WPF app.

Running on Linux or macOS requires the app to be fully UWP or WinUI, including what is currently your WPF shell. Still, once you have started to migrate your WPF app to WinUI (preferably), you can reuse what you migrated on other platforms, while keeping your WPF app that same migrated content with Uno Islands.

**Reply by samsunlu55atakum** [reply]
@jeromelaban hi, when I talk to @martin Zikmund he said there is no way to migrating wpf app to uno platform, You have to rewrite the code. I thought if I use xaml island to use uwp(winUI) in wpf, it will be going well because Uno supports uwp(WinUI). But i don't know how to port my wpf to uno without rewrite xaml and codebehind. Because wpf use system.UI while uno use WinUI. I mean i am stuck. the documentation provide "you have to change some points in wpf to use in UNO".

**Comment by samsunlu55atakum** [comment]
@jeromelaban Can you explain step by step migrating wpf app to uno platform. But I emphasize that I am using xaml island to use uwp(winUI) in wpf. You should think that I am using windowsXamlHost. Will UNO support this migrating? For example like this: 

![Ekran Görüntüsü (20)](https://github.com/unoplatform/uno/assets/143163037/4db92bed-7318-40fe-8556-152878812b71)

**Reply by MartinZikmund** [reply]
Hi, as I mentioned on Discord, if you migrate you need to do the following:

- WPF UI needs to be **manually** migrated to WinUI (Uno) XAML  
- WPF Code behind must be also rewritten **manually**
- Your Islands content can be copied as is into your new Uno app, no changes needed

You don't need `WindowsXamlHost` at all, as that is only to embed content in WPF app. When you are writing a Uno app, you just omit this "host" element and include the intended control as part of the tree. Imagine this as just writing a WinUI app - if you were writing a WinUI app you would also not include any islands - it is already a WinUI app

**Comment by samsunlu55atakum** [comment]
@jeromelaban  hi, when I talk to @Martin Zikmund he said there is no way to migrating wpf app to uno platform, You have to rewrite the code. I thought if I use xaml island to use uwp(winUI) in wpf, it will be going well because Uno supports uwp(WinUI).  But i don't know how to port my wpf to uno without rewrite xaml and codebehind. Because wpf use system.UI while uno use WinUI. I mean i am stuck. the documentation provide "you have to change some points in wpf to use in UNO".

**Reply by jeromelaban** [reply, marked as answer]
> hi, when I talk to @martin Zikmund he said there is no way to migrating wpf app to uno platform,

That is correct, you have to migrate your WPF code to WinUI and then use Uno to make it run on other platforms. 

> But i don't know how to port my wpf to uno without rewrite xaml and codebehind. Because wpf use system.UI while uno use WinUI. I mean i am stuck. the documentation provide "you have to change some points in wpf to use in UNO".

There's no other way around it. You'll have to [adjust your WPF code to be WinUI compatible](https://platform.uno/docs/articles/wpf-migration.html) if you want to make it run on all platforms. In general though, the code is very similar between WinUI and WPF, and it generally straight forward to convert.

> @jeromelaban Can you explain step by step migrating wpf app to uno platform. But I emphasize that I am using xaml island to use uwp(winUI) in wpf. You should think that I am using windowsXamlHost. Will UNO support this migrating? For example like this:

XAML Islands uses WinAppSDK to embed WinUI code in a WPF app. Uno Islands uses Uno to embed WinUI code in a WPF app. Both only run on Windows.

Neither of those Islands technologies will help you run your WPF-specific code on macOS or Linux, but converting your code to WinUI to run in an island will allow you to make a **separate** app that also uses the code that runs in the WPF islands app.

**Reply by samsunlu55atakum** [reply]
@jeromelaban "migrate your wpf code to WinUI" what do you mean, I understand thing is I have to use xaml island. Because there is no other way. so, what do you mean wpf code to winUI?

**Reply by jeromelaban** [reply]
Make sure to [read this article](https://platform.uno/docs/articles/wpf-migration.html) first.

**Reply by samsunlu55atakum** [reply]
the link says that wpf codebase (and xaml) is compatiable %99 with uno. Am I wrong?

**Reply by MartinZikmund** [reply]
Yes it is. But still, you need to do some manual changes, there are things in WPF that don't have the same counterpart or have some different properties in WinUI/Uno

**Reply by darenm** [reply]
This video is a little dated, and speaks more specifically to UWP and Web, but much of the content will show you the areas in your WPF application that need to be addressed when migrating to UWP/WinUI.

[![UnoConf 2021 - Migrating WPF Apps to Web with Daren May](https://img.youtube.com/vi/doomI2bEBno/0.jpg)](https://www.youtube.com/watch?v=doomI2bEBno)
